packages/react-reconciler/src/ReactFiberCompleteWork.js

```diff
@@ -930,112 +1004,160 @@ function completeWork(
-            // Schedule work so we know not to bail out.
-            workInProgress.expirationTime = renderExpirationTime;
-            return workInProgress;
+            // Reset the child fibers to their original state.
+            resetChildFibers(workInProgress, renderExpirationTime);
+
+            // Set up the Suspense Context to force suspense and immediately
+            // rerender the children.
+            pushSuspenseContext(
+              workInProgress,
+              setShallowSuspenseContext(
+                suspenseStackCursor.current,
+                ForceSuspenseFallback,
+              ),
+            );
+            return workInProgress.child;
           }
-        } else {
-          // Append the rendered row to the child list.
-          let rendered = suspenseListState.rendering;
-          if (!suspenseListState.didSuspend) {
-            if (
-              now() > suspenseListState.tailExpiration &&
-              renderExpirationTime > Never
-            ) {
-              // We have now passed our CPU deadline and we'll just give up further
-              // attempts to render the main content and only render fallbacks.
-              // The assumption is that this is usually faster.
-              suspenseListState.didSuspend = true;
-              // Since nothing actually suspended, there will nothing to ping this
-              // to get it started back up to attempt the next item. If we can show
-              // them, then they really have the same priority as this render.
-              // So we'll pick it back up the very next render pass once we've had
-              // an opportunity to yield for paint.
+          // hasSuspendedChildrenAndNewContent could've set didSuspendAlready
+          didSuspendAlready =
+            (workInProgress.effectTag & DidCapture) !== NoEffect;
+        }
+        if (didSuspendAlready) {
+          cutOffTailIfNeeded(renderState);
+        }
+        // Next we're going to render the tail.
+      } else {
+        // Append the rendered row to the child list.
+        if (!didSuspendAlready) {
+          if (
+            now() > renderState.tailExpiration &&
+            renderExpirationTime > Never
+          ) {
+            // We have now passed our CPU deadline and we'll just give up further
+            // attempts to render the main content and only render fallbacks.
+            // The assumption is that this is usually faster.
+            workInProgress.effectTag |= DidCapture;
+            didSuspendAlready = true;
 
-              const nextPriority = renderExpirationTime - 1;
-              workInProgress.expirationTime = workInProgress.childExpirationTime = nextPriority;
-              if (enableSchedulerTracing) {
-                markSpawnedWork(nextPriority);
-              }
-            } else {
-              suspenseListState.didSuspend = isShowingAnyFallbacks(rendered);
+            cutOffTailIfNeeded(renderState);
```

Comment:
This assumes that this didn't suspend but it might have and that would affect whether we keep the rendered row or not. We should probably move the isShowingAnyFallbacks branch up.